### ISSUE TITLE
Can I exec multiple commands with 1 channel?

### ISSUE THREAD

**Issue by TheRadioGuy** [alexcrichton/ssh2-rs#186]

**Comment by crisidev** [comment]
Yes, until you don't disconnect the channel, you can run as many as them you want. If you want to run commands in parallel, it's another story and I think you will need multiple channels.

**Comment by DuckerMan** [comment]
```rust
 pub fn exec(){
let tcp = TcpStream::connect("localhost:22").unwrap();
self.sess.set_tcp_stream(tcp);
self.sess.handshake().unwrap();
self.sess.userauth_password("root","password").unwrap();
assert!(self.sess.authenticated());
let mut ch= self.sess.channel_session().unwrap();
ch.shell().unwrap();
ch.exec(command).unwrap();
ch.exec(command).unwrap();
} 
```

After second command execution, it panics:

```
 thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { code: -39, msg: "Channel can not be reused" }', src/main.rs:35:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by crisidev** [comment]
Than I was very wrong and now that I checked my code, I am creating and closing a new channel per command I execute.. sorry for the wrong info. This is due to how libssh2 works afaik, nothing to do with the rust wapper.

**Comment by DuckerMan** [comment]
So, if I create channel and type `cd directory` then create another channel - will I be in that directory?

**Comment by crisidev** [comment]
I don't think so.. if I may ask, what is your usecase? I am running just commands without having to cd around.. One option would be to run `sh -c "cd directory && mycommand"` in the channel if you need to switch directory.

**Comment by DuckerMan** [comment]
My usecase is kinda ssh-client. So, do I have to create another channel for every command execution? It's no cool to use `&&`

**Comment by crisidev** [comment]
Well if you want an interactive access, you can request a pty and get an interactive terminal. I am on my phone now, but there is an issue already explaining how to do it.. maybe that can help you :)

**Comment by LordFoom** [comment]
For example, I'm writing a little util to ssh onto a server, cd into a directory, rsync a file into that directory, and then extract that file.

Is the only way to do that using &&? That is not very flexible and rather brittle.

I tried opening a channel, cd'ing, opening a new channel, but then it's still in the old directory.

**Comment by wez** [comment]
You need to keep in mind how things work under the covers.  This crate is exposing how the ssh2 protocol works, which is that you can start a process on the remote host and work with its stdio streams.  It is not all that different from considering how eg: the python subprocess class works, it's just wrapped up in a protocol to transport the data over a network link.

With that as your mental model, think about how you would achieve your goal; you can either run a series of separate subprocesses and do something with the results, or start a single remote shell process and send it commands on its stdin stream, one per line, that tell it what to do--in that case you're writing a shell script to the stdin of the remote shell process and consuming its stdout/stderr streams.

**Comment by LordFoom** [comment]
Thank you for your reply, I'm having a look at https://github.com/alexcrichton/ssh2-rs/issues/128 which is helping tremendously.

**Comment by ZiCog** [comment]
Yes.

In this issue report https://github.com/alexcrichton/ssh2-rs/issues/193 there is an example of running two commands in two threads with only a single connection established. The commands are expected to produce output forever and run in parallel.

**Comment by yodaldevoid** [comment]
I'm going to close this issue as I think the question has been answered. If I am wrong, please let me know and I will open this up again.
